src/arithmetic.jl

```diff
@@ -18,63 +22,198 @@ for f in [:inv, :sqrt, :log, :log1p, :exp, :expm1, :atan, :cosh, :sinh]
     @eval Base.$f(x::MagOrRef) = $(Symbol(f, :!))(zero(x), x)
 end
 
-Base.min(x::MagOrRef, y::MagOrRef) = min!(zero(x), x, y)
-Base.max(x::MagOrRef, y::MagOrRef) = max!(zero(x), x, y)
-Base.minmax(x::MagOrRef, y::MagOrRef) = (min(x, y), max(x, y))
+### Arf
+function Base.sign(x::ArfOrRef)
+    isnan(x) && return Arf(NaN) # Follow Julia and return NaN
+    return Arf(sgn(x))
+end
 
-### Arf, Arb and Acb
+Base.abs(x::ArfOrRef) = abs!(zero(x), x)
+Base.:(-)(x::ArfOrRef) = neg!(zero(x), x)
 for (jf, af) in [(:+, :add!), (:-, :sub!), (:*, :mul!), (:/, :div!)]
-    @eval function Base.$jf(x::T, y::T) where {T<:Union{Arf,ArfRef,Arb,ArbRef,Acb,AcbRef}}
-        z = T(prec = max(precision(x), precision(y)))
+    @eval function Base.$jf(x::ArfOrRef, y::Union{ArfOrRef,_BitInteger})
+        z = Arf(prec = _precision((x, y)))
         $af(z, x, y)
-        z
+        return z
     end
 end
-function Base.:(-)(x::T) where {T<:Union{Arf,ArfRef,Arb,ArbRef,Acb,AcbRef}}
-    z = T(prec = precision(x))
-    neg!(z, x)
-    z
+function Base.:+(x::_BitInteger, y::ArfOrRef)
+    z = zero(y)
+    add!(z, y, x)
+    return z
+end
+function Base.:*(x::_BitInteger, y::ArfOrRef)
+    z = zero(y)
+    mul!(z, y, x)
+    return z
 end
-function Base.:(^)(x::T, k::Integer) where {T<:Union{Arb,ArbRef,Acb,AcbRef}}
-    z = T(prec = precision(x))
-    pow!(z, x, convert(UInt, k))
-    z
+function Base.:/(x::_BitUnsigned, y::ArfOrRef)
+    z = zero(y)
+    ui_div!(z, x, y)
+    return z
+end
+function Base.:/(x::_BitSigned, y::ArfOrRef)
+    z = zero(y)
+    si_div!(z, x, y)
+    return z
 end
 
+function Base.sqrt(x::ArfOrRef)
+    y = zero(x)
+    sqrt!(y, x)
+    return y
+end
+function rsqrt(x::ArfOrRef)
+    y = zero(x)
+    rsqrt!(y, x)
+    return y
+end
+function root(x::ArfOrRef, k::Integer)
+    y = zero(x)
+    root!(y, x, convert(UInt, k))
+    return y
+end
+
+### Arb and Acb
+for (jf, af) in [(:+, :add!), (:-, :sub!), (:*, :mul!), (:/, :div!)]
+    @eval Base.$jf(x::ArbOrRef, y::Union{ArbOrRef,ArfOrRef,_BitInteger}) =
+        $af(Arb(prec = _precision((x, y))), x, y)
+    @eval Base.$jf(x::AcbOrRef, y::Union{AcbOrRef,ArbOrRef,_BitInteger}) =
+        $af(Acb(prec = _precision((x, y))), x, y)
+    if jf == :(+) || jf == :(*)
+        @eval Base.$jf(x::Union{ArfOrRef,_BitInteger}, y::ArbOrRef) =
+            $af(Arb(prec = _precision((x, y))), y, x)
+        @eval Base.$jf(x::Union{ArbOrRef,_BitInteger}, y::AcbOrRef) =
+            $af(Acb(prec = _precision((x, y))), y, x)
+    end
+end
+
+Base.:(-)(x::Union{ArbOrRef,AcbOrRef}) = neg!(zero(x), x)
+Base.abs(x::ArbOrRef) = abs!(zero(x), x)
+Base.:(/)(x::_BitUnsigned, y::ArbOrRef) = ui_div!(zero(y), x, y)
+
+# TODO: Should we convert Int to UInt for performance reasons?
+Base.:(^)(x::ArbOrRef, y::Union{ArbOrRef,_BitUnsigned}) =
+    pow!(Arb(prec = _precision((x, y))), x, y)
+Base.:(^)(x::AcbOrRef, y::Union{AcbOrRef,ArbOrRef,_BitInteger}) =
+    pow!(Acb(prec = _precision((x, y))), x, y)
```

Comment:
since we're writing these methods directly I'd
```julia
Base.:(^)(x::ArbOrRef, y::ArbOrRef) = 
    pow!(Arb(prec = _precision((x, y))), x, y)
function Base.:(^)(x::ArbOrRef, y::_BitInteger)
    x, n = (y >= 0 ? (x, y) : (inv(x), abs(y)))
    return pow!(Arb(prec = precision(x)), convert(UInt, n))
end
Base.:(^)(x::AcbOrRef, y::Union{AcbOrRef,ArbOrRef,_BitInteger}) =
    pow!(Acb(prec = _precision((x, y))), x, y)

Reply:
I a slightly different version of this that avoids one allocation by using `inv!`. I don't think we have to extend `Base.literal_pow` directly, the reason the first definition is not used in your example is that `-10` is not a `UInt` so it falls back to the default power by squaring. It works now
```julia
julia> Arb(2)^-10
0.00097656250000000000000000000000000000000000000000000000000000000000000000000000
```
We could possibly extend `Base.literal_pow` to use `Arblib.sqr` for `^2`. It seems to be about 10% faster actually when I try it.

Reply:
I extended `Base.literal_pow` for squaring now! I realized that `arb_pow_ui` actually starts by converting the `ui` to `fmpz` so this always require an extra allocation even in the cases which are later optimized, which are exponents `-2, -1, 0, 1, 2`. Adding a literal pow for all of these would avoid one allocation, at the moment `-1` and `2` are implemented.